File: hooks/Config_ProactiveEvaluation/hook/src/main/java/com/awscommunity/config/proactiveeval/BaseHookHandlerStd.java
package com.awscommunity.config.proactiveeval;

import org.apache.commons.lang3.exception.ExceptionUtils;

import software.amazon.awssdk.services.config.model.InvalidParameterValueException;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.HandlerErrorCode;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.ProgressEvent;
import software.amazon.cloudformation.proxy.hook.HookContext;
import software.amazon.cloudformation.proxy.hook.HookHandlerRequest;
import software.amazon.cloudformation.proxy.hook.targetmodel.HookTargetModel;

/**
 * Methods that are common to both pre-create and pre-update handlers for this
 * hook.
 */
public abstract class BaseHookHandlerStd
    extends BaseHookHandler<CallbackContext, TypeConfigurationModel> {

  private static final String insufficientDataOutcomeDefault = "PASS";

  /**
   * Run operations common to pre-create and pre-update phases for this hook.
   *
   * @param proxy             AmazonWebServicesClientProxy
   * @param request           HookHandlerRequest
   * @param callbackContext   CallbackContext
   * @param logger            Logger
   * @param typeConfiguration TypeConfigurationModel
   * @return ProgressEvent
   */
  protected ProgressEvent<HookTargetModel, CallbackContext> preCreatePreUpdateOperations(
      final AmazonWebServicesClientProxy proxy,
      final HookHandlerRequest request,
      final CallbackContext callbackContext,
      final Logger logger,
      final TypeConfigurationModel typeConfiguration) {

    final HookContext hookContext = request.getHookContext();
    final String resourceType = hookContext.getTargetName();
    final String resourceId = hookContext.getTargetLogicalId();
    final String resourceConfiguration = hookContext
        .getTargetModel()
        .getTargetModelAsJSONObject()
        .get("resourceProperties")
        .toString();

    try {

      ProactiveComplianceHandler proactiveComplianceHandler = ProactiveComplianceHandler.builder()
          .resourceType(resourceType)
          .resourceId(resourceId)
          .resourceConfiguration(resourceConfiguration)
          .awsClientProxy(proxy)
          .outcomeForComplianceTypeInsufficientData(getInsufficientDataOutcome(typeConfiguration))
          .configClient(ClientBuilder.getConfigClient())
          .build();

      String evaluationId = proactiveComplianceHandler.startResourceEvalution();
      return proactiveComplianceHandler.assessCompliance(evaluationId);

    } catch (final InvalidParameterValueException ipve) {
      final String messageForInvalidParameterValueException = ipve.getMessage()
          + "  Check that you have specified all the required parameters"
          + " for the resource type you are describing"
          + " in your CloudFormation template, and parameters' key(s)/value(s) are correct.";
      logger.log(messageForInvalidParameterValueException);

      return ProgressEvent.<HookTargetModel, CallbackContext>builder()
          .status(OperationStatus.FAILED)
          .message(messageForInvalidParameterValueException)
          .errorCode(HandlerErrorCode.InvalidRequest)
          .build();

    } catch (final Throwable throwable) {
      logger.log(ExceptionUtils.getStackTrace(throwable));

      String messageFromThrowable = "A handler internal failure error has occurred.";
      if (throwable.getMessage() != null) {
        messageFromThrowable += String.format(" %s", throwable.getMessage());
      }

      return ProgressEvent.<HookTargetModel, CallbackContext>builder()
          .status(OperationStatus.FAILED)
          .message(messageFromThrowable)
          .errorCode(HandlerErrorCode.HandlerInternalFailure)
          .build();
    }
  }

  /**
   * Return the default value for the InsufficientDataOutcome behavior if not set
   * in the type configuration; otherwise, return the value set in the type
   * configuration.
   *
   * @param typeConfiguration TypeConfigurationModel
   * @return String
   */
  private String getInsufficientDataOutcome(final TypeConfigurationModel typeConfiguration) {
    final String outcome = typeConfiguration.getOutcomeForComplianceTypeInsufficientData();

    return (outcome == null || outcome.isEmpty()) ? insufficientDataOutcomeDefault : outcome;
  }
}

File: hooks/S3_AccessControl/src/main/kotlin/com/awscommunity/s3/accesscontrol/BaseHookHandlerStd.kt
package com.awscommunity.s3.accesscontrol

import com.awscommunity.s3.accesscontrol.model.aws.s3.bucket.AwsS3BucketTargetModel
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.OperationStatus
import software.amazon.cloudformation.proxy.ProgressEvent
import software.amazon.cloudformation.proxy.hook.HookHandlerRequest
import software.amazon.cloudformation.proxy.hook.targetmodel.HookTargetModel

/**
 * This class is used by both pre-create and pre-update handlers for this hook,
 * and it contains common business logic for validation of target(s).
 */
abstract class BaseHookHandlerStd : BaseHookHandler<CallbackContext?, TypeConfigurationModel>() {

    /**
     * Common entry point for pre-create and pre-update validation operations.
     *
     * @param proxy AmazonWebServicesClientProxy
     * @param request HookHandlerRequest
     * @param callbackContext CallbackContext
     * @param logger Logger
     * @param typeConfiguration TypeConfigurationModel
     */
    fun handlePreCreatePreUpdateRequests(
        proxy: AmazonWebServicesClientProxy,
        request: HookHandlerRequest,
        callbackContext: CallbackContext?,
        logger: Logger,
        typeConfiguration: TypeConfigurationModel?
    ): ProgressEvent<HookTargetModel, CallbackContext?> {
        val targetModel = request.hookContext.getTargetModel(AwsS3BucketTargetModel::class.java)
        val resourceProperties = targetModel.resourceProperties
        val accessControl: String? = resourceProperties.accessControl
        if (accessControl != null && accessControl != "Private") {
            return ProgressEvent.builder<HookTargetModel, CallbackContext?>()
                .status(OperationStatus.FAILED)
                .errorCode(HandlerErrorCode.NonCompliant)
                .message("The legacy AccessControl property is present in the resource's configuration, and it is not set to Private.  Remove the legacy AccessControl property, or set it to Private.")
                .build()
        }

        return ProgressEvent.builder<HookTargetModel, CallbackContext?>()
            .status(OperationStatus.SUCCESS)
            .message("The legacy AccessControl property is either not present in the resource's configuration, or it is set to Private.")
            .build()
    }
}
